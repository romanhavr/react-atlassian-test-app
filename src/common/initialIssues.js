export const initialIssues = [
  {
    id: 1,
    content: {
      id: 1,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 2,
    content: {
      id: 2,
      issue: '2-nd',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 2,
      labelIds: [2]
    }
  },
  {
    id: 3,
    content: {
      id: 3,
      issue: '1-st copy',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 4,
    content: {
      id: 4,
      issue: '3-rd',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [2, 3]
    }
  },
  {
    id: 5,
    content: {
      id: 5,
      issue: '1-st another copy',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 3,
      labelIds: [1, 3]
    }
  },
  {
    id: 6,
    content: {
      id: 6,
      issue: '1-st',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [1, 3]
    }
  },
  {
    id: 7,
    content: {
      id: 7,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 2,
      labelIds: [1, 3]
    }
  }
]