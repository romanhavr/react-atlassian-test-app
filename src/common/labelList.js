export const labels = [
    { id: 1, name: 'label1', label: 'Label 1' },
    { id: 2, name: 'anotherLabel', label: 'Another Label' },
    { id: 3, name: 'newLabel', label: 'New Label' },
]