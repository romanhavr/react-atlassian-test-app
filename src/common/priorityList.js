export const priorityList = [
    { level: 1, label: 'High', value: 'high' },
    { level: 2, label: 'Medium', value: 'medium' },
    { level: 3, label: 'Low', value: 'low' }
]