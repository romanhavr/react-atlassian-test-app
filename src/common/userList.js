// @flow

import type { User } from '../interfaces/interfaces';

export const userList: User[] = [
    {
        id: 1,
        displayName: 'Admin',
        avatar: 'https://pickaface.net/gallery/avatar/unr_edwinavator_170412_0152_yoidv.png'
    },
    {
        id: 2,
        displayName: 'Elvis',
        avatar: 'https://pickaface.net/gallery/avatar/Nirvana53183d8f601d7.png'
    },
    {
        id: 3,
        displayName: 'Roling Stone',
        avatar: 'https://tradiesinbusiness.com.au/wp-content/uploads/2019/09/avatar-jongen-voorbeeld-1.jpg'
    },
    {
        id: 4,
        displayName: 'John Smith',
        avatar: 'https://apollodesign.nl/wp-content/uploads/2015/08/avatar-casper-tekening.png'
    },
]