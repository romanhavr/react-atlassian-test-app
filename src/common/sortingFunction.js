// @flow

import type { TableItem, SortingOptions, SortingCriteria } from '../interfaces/interfaces';
import { priorityList } from '../common/priorityList';
import { userList } from '../common/userList';

export function sortingFunc(a: TableItem, b: TableItem, sortingBy: SortingOptions, i: number): number {
        if (comparing(a, b, sortingBy[i].value, 'more')) {
            return 1
        } else if (comparing(a, b, sortingBy[i].value, 'less')) {
            return -1
        } else if (comparing(a, b, sortingBy[i].value, 'eq') && i < sortingBy.length - 1) {
            return sortingFunc(a, b, sortingBy, ++i)
        } else
            return 0
};

function comparing (a: TableItem, b: TableItem, sortingValue: string, compareIndex: string): boolean {
    const newObject: SortingCriteria = sortingCriteria (a.content[sortingValue], b.content[sortingValue], sortingValue)
    switch (compareIndex) {
        case 'more' :
            return newObject.newObj1 > newObject.newObj2;
        case 'eq' :
            return newObject.newObj1 === newObject.newObj2;
        case 'less' :
            return newObject.newObj1 < newObject.newObj2;
        default:
            return newObject.newObj1 > newObject.newObj2;
    }
}

function sortingCriteria (obj1, obj2, sortingValue: string): SortingCriteria {
    switch(sortingValue) {
        case 'issue' :
            return { newObj1: obj1, newObj2: obj2};
        case 'assignee' :
            const userById1 = userList.find(el => el.id === obj1);
            const userById2 = userList.find(el => el.id === obj2);
            if (userById1 && userById2) return { 
                newObj1: userById1.displayName,
                newObj2: userById2.displayName
            };
            break;
        case 'priority' :
            const priorityByValue1 = priorityList.find(el => el.value === obj1.value);
            const priorityByValue2 = priorityList.find(el => el.value === obj2.value);
            if (priorityByValue1 && priorityByValue2) return {
                newObj1: priorityByValue1.level,
                newObj2: priorityByValue2.level
            };
            break;
        default:
            return { newObj1: obj1, newObj2: obj2};
    }
}