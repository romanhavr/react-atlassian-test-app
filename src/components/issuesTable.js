// @flow

import React from 'react';
import { connect } from 'react-redux';
import '../styles/modalWindow.css';
import { chooseItem, editIssue, removeIssue } from '../store/actions';
import TableTree from '@atlaskit/table-tree';
import Avatar from '@atlaskit/avatar';
import Badge from '@atlaskit/badge';
import ModalDialog, { ModalTransition } from '@atlaskit/modal-dialog';
import Form from '@atlaskit/form';
import CreateEditForm from './create-edit-form';
import { userList } from '../common/userList';
import { labels as labelList } from '../common/labelList';
import { priorityList } from '../common/priorityList';
import { sortingFunc } from '../common/sortingFunction';
import PriorityMajorIcon from '@atlaskit/icon-priority/glyph/priority-major';
import PriorityMediumIcon from '@atlaskit/icon-priority/glyph/priority-medium';
import PriorityMinorIcon from '@atlaskit/icon-priority/glyph/priority-minor';
import TrashIcon from '@atlaskit/icon/glyph/trash';
import type { Item,
            User,
            Priority,
            SortingOptions,
            TableItem,
            ActionButton,
            EditFormData,
            EditFormActionData
        } from '../interfaces/interfaces';

type Props = {
    storeTableData: Item[],
    chosenItem: Item,
    sortingBy: SortingOptions,
    chooseItemClick: any,
    editIssue: any,
    itemRemove: any
}

type State = {
    isOpen: boolean,
    tableData : Item[]
};

export class IssuesTable extends React.Component<Props, State> {

    chooseItemClick = this.props.chooseItemClick;
    editIssue = this.props.editIssue;
    itemRemove = this.props.itemRemove;
    state: State = {
        isOpen: !!this.props.chosenItem,
        tableData : this.props.storeTableData
    };
    keyCounter = 0;

    itemClick = (item: Item )=> {
        this.chooseItemClick(item);
        this.setState({ isOpen: true })
    }

    close = () => {
        this.chooseItemClick(null);
        this.setState({ isOpen: false })
    };

    onFormSubmit = (data: EditFormData) => {
        this.editIssue({...data, id: this.props.chosenItem.id});
        setTimeout(() => this.setState({
            tableData: this.props.storeTableData
        }), 0);
        this.close();
    }

    removeItem = (item: Item) => {
        this.itemRemove(item);
        setTimeout(() => this.setState({
            tableData: this.props.storeTableData
        }), 0);
    }

    render() {
        const { isOpen/*(:boolean)*/ } = this.state;

        const sortedTableData: TableItem[] = this.state.tableData.map(issue => {
            return {
                id: issue.id,
                content: issue
            }
        });

        if (this.props.sortingBy && this.props.sortingBy.length) {
            sortedTableData.sort( (a, b) => sortingFunc(a, b, this.props.sortingBy, 0))
        }

        const actionButtons: ActionButton[] = [
            { text: 'Save changes', type: 'submit' },
            { text: 'Cancel', onClick: this.close },
        ];
        const priorityIcon = (priorityLabel: ?string) => {
            switch(priorityLabel) {
                case 'High' :
                    return <PriorityMajorIcon size="medium" />;
                case 'Medium' :
                    return <PriorityMediumIcon size="medium" />;
                case 'Low' :
                    return <PriorityMinorIcon size="medium" />;
                default:
                    return ''
            }
        }

        const issue = (item: Item) => <span onClick={() => this.itemClick(item)} className="summary">
                                    {item.issue}
                                </span>;
        const assignee = ({ assignee }: Item) => { const user: ?User = userList.find( ({ id }) => id === assignee);
                                if (user) return (<span> <Avatar
                                                    size="xsmall"
                                                    src={user.avatar}
                                                />
                                                {user.displayName}
                                            </span>)};
        const labels = (item: Item) => item.labelIds.map( id => {
                                    this.keyCounter++;
                                    return (
                                        <Badge key={this.keyCounter}>
                                            <strong>
                                                {labelList.filter( label => id === label.id)[0].label.toUpperCase()}
                                            </strong>
                                        </Badge>
                                    )}
                                )                                
        const priority = ({priority}: Item) => {const findItemLabel: ?Priority = priorityList.find(({level}) => level === priority);
                                    const itemLabel: ?string = findItemLabel ? findItemLabel.label : null;
                                    return (<span>
                                        {priorityIcon(itemLabel)}
                                        {itemLabel}
                                    </span>)};
        const removing = (item: Item) => <span className="trash-icon" onClick={() => this.removeItem(item)}><TrashIcon /></span>

        return (
            <React.Fragment>
                <TableTree
                    headers={['Issue', 'Assignee', 'Labels', 'Priority', 'Trash']}
                    columns={[issue, assignee, labels, priority, removing]}
                    columnWidths={['200px', '170px', '200px', '150px', '80px']}
                    items={sortedTableData}
                />
                <ModalTransition>
                    {isOpen && (
                        <ModalDialog
                            actions={actionButtons}
                            onClose={this.close}
                            heading="Edit issue"
                            components={{
                                Container: ({ children }) => (
                                    <Form onSubmit={this.onFormSubmit}>
                                        {({ formProps }) =>( 
                                            <form className="form" {...formProps}>
                                                {children}
                                            </form>
                                        )}
                                    </Form>
                                )
                            }}
                        >
                            <CreateEditForm />
                        </ModalDialog>)
                    }
                </ModalTransition>
            </React.Fragment>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        storeTableData: state.issues,
        chosenItem: state.ui.chosenItem,
        sortingBy: state.ui.sortBy
    }
}

const mapDispatchToProps = dispatch => {
    return {
        chooseItemClick: (item: Item) => dispatch(chooseItem(item)),
        editIssue: (data: EditFormActionData) => dispatch(editIssue(data)),
        itemRemove: (item: Item) => dispatch(removeIssue(item))
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(IssuesTable);