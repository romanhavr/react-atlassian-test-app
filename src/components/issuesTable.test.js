// @flow

import { shallow, mount } from 'enzyme';
import React from 'react';
import { IssuesTable } from './issuesTable';
import { initialIssues } from '../common/initialIssues';
import { Provider } from "react-redux";
import TableTree from '@atlaskit/table-tree';
import configureMockStore from "redux-mock-store";
import type { Item, TableItem } from '../interfaces/interfaces';

const mockStore = configureMockStore();

const chosenItem: Item = {
    id: 1572444816584,
    issue: "1-st copy",
    priority: 1,
    assignee: 1,
    labelIds: [1]
}

const tableData: TableItem[] = initialIssues.map(issue => ({id: issue.id, content: issue}))

const initialState = {
    storeTableData: initialIssues,
    tableData,
    chosenItem,
    sortingBy: []
}
const initialDispatches = {
    chooseItemClick: jest.fn(),
    editIssue: jest.fn(),
    editAssigneeInline: jest.fn(),
    editIssueInline: jest.fn(),
    editLabelsInline: jest.fn(),
    editPriorityInline: jest.fn(),
    itemRemove: jest.fn()
}

const initialStateForMount = {...initialState, ...initialDispatches}

const store = mockStore({ ui: { chosenItem } } );

const shallowWrapper = shallow(<IssuesTable {...initialStateForMount}/>);

const mountWrapper = mount(
    <Provider store={store}>
        <IssuesTable {...initialStateForMount} />
    </Provider>
);

const sortedTableData: TableItem[] = initialIssues.map(issue => {
    return {
        id: issue.id,
        content: issue
    }
});

describe('Issues Table testing...', () => {

    it('IssueTable should include TableTree (using "shallow")', () => {
        expect(shallowWrapper.find('TableTree')).toExist();
    })

    it('IssueTable should include spans (using "mount")', () => {
        expect(mountWrapper.find('span')).toExist();
    })

    const form = mountWrapper.find('form');
    const inlineEditIssue = mountWrapper.find('[data-test="inline-edit-issue"]');
    const inlineEditAssignee = mountWrapper.find('[data-test="inline-edit-assignee"]');
    const inlineEditLabels = mountWrapper.find('[data-test="inline-edit-labels"]');
    const inlineEditPriority = mountWrapper.find('[data-test="inline-edit-priority"]');

    it('IssueTable should include form (using "mount")', () => {
        expect(form).toExist();

        const onSubmitFn = jest.fn();
        const onCloseFn = jest.fn();
        const shallowWrapperForm = shallow(<form onSubmit={onSubmitFn} onClose={onCloseFn} />);
        const form1 = shallowWrapperForm.find('form')
        expect(form1).toExist();

        form1.simulate('submit', chosenItem);
        expect(onSubmitFn).toHaveBeenCalledTimes(1);
        expect(onSubmitFn).toHaveBeenLastCalledWith(chosenItem);

        form1.simulate('close');
        expect(onCloseFn).toHaveBeenCalledTimes(1);
      });

    it('IssueTable should include SNAPSHOT', () => {
        const tree = shallow(
        <TableTree
                headers={['Issue', 'Assignee', 'Labels', 'Priority', 'Trash']}
                columns={['issue', 'assignee', 'labels', 'priority', 'removing']}
                columnWidths={['200px', '170px', '200px', '150px', '80px']}
                items={sortedTableData}
            />
            );
        
        expect(tree).toMatchSnapshot();
    })

    it('IssueTable should include inline edit elements and they should dispatch actions', () => {
        expect(inlineEditIssue).toExist();
        expect(inlineEditAssignee).toExist();
        expect(inlineEditLabels).toExist();
        expect(inlineEditPriority).toExist();
        
        inlineEditIssue.at(1).simulate('submit');
        inlineEditAssignee.at(1).simulate('submit');
        inlineEditLabels.at(1).simulate('submit');
        inlineEditPriority.at(1).simulate('submit');
        expect(initialDispatches.editIssueInline).toHaveBeenCalled();
        expect(initialDispatches.editAssigneeInline).toHaveBeenCalled();
        expect(initialDispatches.editLabelsInline).toHaveBeenCalled();
        expect(initialDispatches.editPriorityInline).toHaveBeenCalled();
    })
})