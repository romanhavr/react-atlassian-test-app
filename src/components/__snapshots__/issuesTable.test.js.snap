// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issues Table testing... IssueTable should include SNAPSHOT 1`] = `
<styled__TableTreeContainer
  aria-readonly={true}
  role="treegrid"
>
  <Headers>
    <CellWithColumnWidth
      columnIndex={0}
      key="0"
      width="200px"
    >
      Issue
    </CellWithColumnWidth>
    <CellWithColumnWidth
      columnIndex={1}
      key="1"
      width="200px"
    >
      Assignee
    </CellWithColumnWidth>
    <CellWithColumnWidth
      columnIndex={2}
      key="2"
      width="250px"
    >
      Labels
    </CellWithColumnWidth>
    <CellWithColumnWidth
      columnIndex={3}
      key="3"
      width="150px"
    >
      Priority
    </CellWithColumnWidth>
  </Headers>
  <Rows
    items={
      Array [
        Object {
          "content": Object {
            "assignee": 1,
            "id": 1572444816582,
            "issue": "1-st",
            "labelIds": Array [
              1,
            ],
            "priority": 2,
          },
          "id": 1572444816582,
        },
        Object {
          "content": Object {
            "assignee": 2,
            "id": 1572444816583,
            "issue": "2-nd",
            "labelIds": Array [
              2,
            ],
            "priority": 3,
          },
          "id": 1572444816583,
        },
        Object {
          "content": Object {
            "assignee": 1,
            "id": 1572444816584,
            "issue": "1-st copy",
            "labelIds": Array [
              1,
            ],
            "priority": 1,
          },
          "id": 1572444816584,
        },
        Object {
          "content": Object {
            "assignee": 4,
            "id": 1572444816585,
            "issue": "3-rd",
            "labelIds": Array [
              2,
              3,
            ],
            "priority": 1,
          },
          "id": 1572444816585,
        },
        Object {
          "content": Object {
            "assignee": 3,
            "id": 1572444816586,
            "issue": "1-st another copy",
            "labelIds": Array [
              1,
              3,
            ],
            "priority": 3,
          },
          "id": 1572444816586,
        },
        Object {
          "content": Object {
            "assignee": 4,
            "id": 1572444816587,
            "issue": "1-st",
            "labelIds": Array [
              1,
              3,
            ],
            "priority": 1,
          },
          "id": 1572444816587,
        },
        Object {
          "content": Object {
            "assignee": 2,
            "id": 1572444816588,
            "issue": "1-st",
            "labelIds": Array [
              1,
              3,
            ],
            "priority": 2,
          },
          "id": 1572444816588,
        },
      ]
    }
    render={[Function]}
  />
</styled__TableTreeContainer>
`;
